pkg/client/cloudevents/source_client.go

```diff
@@ -0,0 +1,132 @@
+package cloudevents
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+
+	"github.com/openshift-online/maestro/pkg/api"
+	"github.com/openshift-online/maestro/pkg/logger"
+	"github.com/openshift-online/maestro/pkg/services"
+	cegeneric "open-cluster-management.io/api/cloudevents/generic"
+	ceoptions "open-cluster-management.io/api/cloudevents/generic/options"
+	cetypes "open-cluster-management.io/api/cloudevents/generic/types"
+)
+
+type SourceClient interface {
+	OnCreate(ctx context.Context, id string) error
+	OnUpdate(ctx context.Context, id string) error
+	OnDelete(ctx context.Context, id string) error
+}
+
+type SourceClientImpl struct {
+	Codec                  cegeneric.Codec[*api.Resource]
+	CloudEventSourceClient *cegeneric.CloudEventSourceClient[*api.Resource]
+	ResourceService        services.ResourceService
+}
+
+func NewSourceClient(sourceOptions *ceoptions.CloudEventsSourceOptions, resourceService services.ResourceService) (SourceClient, error) {
+	ctx := context.Background()
+	codec := &Codec{}
+	ceSourceClient, err := cegeneric.NewCloudEventSourceClient[*api.Resource](ctx, sourceOptions,
+		resourceService, ResourceStatusHashGetter, codec)
+	if err != nil {
+		return nil, err
+	}
+
+	logger := logger.NewOCMLogger(ctx)
+	go func() {
+		if err := ceSourceClient.Subscribe(ctx, func(action cetypes.ResourceAction, resource *api.Resource) error {
+			logger.Infof("received action %s for resource %s", action, resource.ID)
+			_, err := resourceService.UpdateStatus(ctx, resource)
```

Comment:
we may have a TODO handle resource deleted status here?

Reply:
Great suggestion! I'm uncertain about its ideal placement—perhaps not significant enough for a standalone readme file. I opted to incorporate the delete flow within the code comments: https://github.com/openshift-online/maestro/pull/12/commits/5809e835573cb3eadf86d9ee6f7c3ced68f2672a